README.markdown
Mission
=======

[![Join the chat at https://gitter.im/fisher/gen_smtp](https://badges.gitter.im/fisher/gen_smtp.svg)](https://gitter.im/fisher/gen_smtp?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

Provide a generic Erlang SMTP server framework that can be extended via
callback modules in the OTP style. A pure Erlang SMTP client is also included.
The goal is to make it easy to send and receive email in Erlang without the
hassle of POP/IMAP. This is *not* a complete mailserver - although it includes
most of the parts you'd need to build one.

The SMTP server/client supports PLAIN, LOGIN, CRAM-MD5 authentication as well
as STARTTLS and SSL (port 465).

Also included is a MIME encoder/decoder, sorta according to RFC204{5,6,7}.

IPv6 is also supported (at least serverside).

I (Vagabond) have had a simple gen_smtp based SMTP server receiving and parsing
copies of all my email for several months and its been able to handle over 100
thousand emails without leaking any RAM or crashing the erlang virtual machine.

Current Participants
====================

+ Andrew Thompson (andrew AT hijacked.us)
+ Jack Danger Canty (code AT jackcanty.com)
+ Micah Warren (micahw AT lordnull.com)

Who is using it?
================

+ gen_smtp is used to provide the email functionality of [OpenACD](http://github.com/Vagabond/OpenACD)
+ gen_smtp will be used as both the SMTP server and SMTP client for [Zotonic](http://zotonic.com) as of version 0.7
+ [Chicago Boss](http://www.chicagoboss.org/) uses gen_smtp for its mail API.
+ [Gmailbox](https://www.gmailbox.org) uses gen_smtp to provide a free email forwarding service.

If you'd like to share your usage of gen_smtp, please contact me.

Client Example
==============

Here's an example usage of the client:

<pre>
gen_smtp_client:send({"whatever@test.com", ["andrew@hijacked.us"],
 "Subject: testing\r\nFrom: Andrew Thompson <andrew@hijacked.us>\r\nTo: Some Dude <foo@bar.com>\r\n\r\nThis is the email body"},
  [{relay, "smtp.gmail.com"}, {username, "me@gmail.com"}, {password, "mypassword"}]).
</pre>

The From and To addresses will be wrapped in &lt;&gt;s if they aren't already,
TLS will be auto-negotiated if available (unless you pass `{tls, never}`) and
authentication will by attempted by default since a username/password were
specified (`{auth, never}` overrides this).

If you want to mandate tls or auth, you can pass `{tls, always}` or `{auth,
always}` as one of the options. You can specify an alternate port with `{port,
2525}` (default is 25) or you can indicate that the server is listening for SSL
connections using `{ssl, true}` (port defaults to 465 with this option).

DKIM signing of outgoing emails
-------------------------------

You may wish to configure DKIM signing [RFC6376](http://tools.ietf.org/html/rfc6376) of outgoing emails
for better security. To do that you need public and private RSA keys, which can be generated by
following commands:
```bash
openssl genrsa -out private-key.pem 1024
openssl rsa -in private-key.pem -out public-key.pem -pubout
```
To send DKIM-signed email:
```erlang
{ok, PrivKey} = file:read_file("private-key.pem"),
DKIMOptions = [
    {s, <<"foo.bar">>},
    {d, <<"example.com">>},
	{private_key, {pem_plain, PrivKey}}]}
    %{private_key, {pem_encrypted, EncryptedPrivKey, "password"}}
],
SignedMailBody = \
 mimemail:encode({<<"text">>, <<"plain">>,
                  [{<<"Subject">>, <<"DKIM testing">>},
                   {<<"From">>, <<"Andrew Thompson <andrew@hijacked.us>">>},
                   {<<"To">>, <<"Some Dude <foo@bar.com>">>}],
                  [],
                  <<"This is the email body">>},
                  [{dkim, DKIMOptions}]),
gen_smtp_client:send({"whatever@example.com", ["andrew@hijacked.us"], SignedMailBody}, []).
```
Don't forget to put your public key to `foo.bar._domainkey.example.com` TXT DNS record as smth like
```
v=DKIM1; g=*; k=rsa; p=MIGfMA0GCSqGSIb3DQEBA......
```
See RFC6376 for more details.

Server Example
==============

gen_smtp ships with a simple callback server example, smtp_server_example. To start the SMTP server with this as the callback module, issue the following command:

<pre>
gen_smtp_server:start(smtp_server_example).
gen_smtp_server starting at nonode@nohost
listening on {0,0,0,0}:2525 via tcp
{ok,<0.33.0>}
</pre>

By default it listens on 0.0.0.0 port 2525. You can telnet to it and test it:

<pre>
^andrew@orz-dashes:: telnet localhost 2525                                                      [~]
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
220 localhost ESMTP smtp_server_example
EHLO example.com
250-orz-dashes
250-SIZE 10485670
250-8BITMIME
250-PIPELINING
250 WTF
MAIL FROM: andrew@hijacked.us
250 sender Ok
RCPT TO: andrew@hijacked.us
250 recipient Ok
DATA
354 enter mail, end with line containing only '.'
Good evening gentlemen, all your base are belong to us.
.
250 queued as #Ref<0.0.0.47>
QUIT
221 Bye
Connection closed by foreign host.
</pre>

You can configure the server in general, each SMTP session, and the callback module, for example:

<pre>
gen_smtp_server:start(smtp_server_example, [[{sessionoptions, [{allow_bare_newlines, fix}, {callbackoptions, [{parse, true}]}]}]]).
</pre>

This configures the session to fix bare newlines (other options are 'strip', 'true' and 'false', false rejects emails with bare newlines, true passes them through unmodified and strip removes them) and tells the callback module to run the MIME decoder on the email once its been received. The example callback module also supports the following options; relay - whether to relay email on, auth - whether to do SMTP authentication and parse - whether to invoke the MIME parser. The example callback module is included mainly as an example and are not intended for serious usage. You could easily create your own callback options.

You can also start multiple SMTP listeners at once by passing more than 1 configuration:

<pre>
gen_smtp_server:start(smtp_server_example, [[], [{protocol, ssl}, {port, 1465}]]).
</pre>

This starts 2 listeners, one with the default config, and one with the default config except that its running in SSL mode on port 1465.

You can connect and test this using the gen_smtp_client via something like:

<pre>
gen_smtp_client:send({"whatever@test.com", ["andrew@hijacked.us"], "Subject: testing\r\nFrom: Andrew Thompson \r\nTo: Some Dude \r\n\r\nThis is the email body"}, [{relay, "localhost"}, {port, 1465}, {ssl, true}]).
</pre>

If you want to listen on IPv6, you can use the {family, inet6} and {address, "::"} options to enable listening on IPv6.

Live Instance
=============

If you want, you can connect to a live instance at mail.cataclysm-software.net (its also the MX record for cataclysm-software.net). Its listening on both IPv4 and v6.


debian/changelog
erlang-gen-smtp (0.9.0.1) unstable; urgency=low

  * oops, wrong dirname
  * stop if yrl compilation fails
  * supress erl console stdout in yrl comp
  * build dep on erlang-parsetools (for yecc)
  * forgot to add some files
  * typo in build-deps
  * satisfy edoc
  * non-functional: small style fix
  * too long lines reduced
  * added dependency on ebin to yrl
  * provided tests are passed
  * added yecc compilation

 -- Serge Rybalchenko <fisher@heim.in.ua>  Tue, 29 Mar 2016 15:37:31 +0300

erlang-gen-smtp (0.9.0) unstable; urgency=medium

  * Initial release.

 -- Serge Rybalchenko <fisher@heim.in.ua>  Mon, 28 Mar 2016 15:31:05 +0200


test/gen_smtp_util_test.erl
-module(gen_smtp_util_test).

-include_lib("eunit/include/eunit.hrl").

test_test() ->
    smtp_util:parse_rfc822_addresses("foo bar").

parse_rfc822_addresses_test_() ->
    [
     {"Empty address list",
      fun() ->
              ?assertEqual({ok, []}, smtp_util:parse_rfc822_addresses(<<>>)),
              ?assertEqual({ok, []},
                           smtp_util:parse_rfc822_addresses(<<"   ">>)),
              ?assertEqual({ok, []},
                           smtp_util:parse_rfc822_addresses(<<" \r\n\t  ">>)),
              ?assertEqual({ok, []},
                           smtp_util:parse_rfc822_addresses(<<10>>)),
              ?assertEqual({ok, []},
                           smtp_util:parse_rfc822_addresses(<<13,10>>)),
              ?assertEqual({ok, []},
                           smtp_util:parse_rfc822_addresses(<<10,13>>)),
              ?assertEqual({ok, []},
                           smtp_util:parse_rfc822_addresses(<<13>>))
      end},

     {"Single addresses",
      fun() ->
              ?assertEqual({ok, [{undefined, "john@doe.com"}]},
                           smtp_util:parse_rfc822_addresses(<<"john@doe.com">>)),
              ?assertEqual(
                 {ok, [{"Fræderik Hølljen", "me@example.com"}]},
                 smtp_util:parse_rfc822_addresses(
                   <<"Fræderik Hølljen <me@example.com>">>)),
              ?assertEqual(
                 {ok, [{undefined, "john@doe.com"}]},
                 smtp_util:parse_rfc822_addresses(<<"<john@doe.com>">>)),
              ?assertEqual(
                 {ok, [{"John", "john@doe.com"}]},
                 smtp_util:parse_rfc822_addresses(<<"John <john@doe.com>">>)),
              ?assertEqual(
                 {ok, [{"John Doe", "john@doe.com"}]},
                 smtp_util:parse_rfc822_addresses(
                   <<"John Doe <john@doe.com>">>)),
              ?assertEqual(
                 {ok, [{"John Doe", "john@doe.com"}]},
                 smtp_util:parse_rfc822_addresses(
                   <<"\"John Doe\" <john@doe.com>">>)),
              ?assertEqual(
                 {ok, [{"John \"Mighty\" Doe", "john@doe.com"}]},
                 smtp_util:parse_rfc822_addresses(
                   <<"\"John \\\"Mighty\\\" Doe\" <john@doe.com>">>))
      end},

     {"Multiple addresses",
      fun() ->
              ?assertEqual(
                 {ok, [{undefined, "a@a.com"}, {undefined, "b@b.com"}]},
                 smtp_util:parse_rfc822_addresses(<<"a@a.com,b@b.com">>)),
              ?assertEqual(
                 {ok, [{undefined, "a,a@a.com"}, {undefined, "b@b.com"}]},
                 smtp_util:parse_rfc822_addresses(<<"<a,a@a.com>,b@b.com">>)),
              ?assertEqual(
                 {ok, [{"Jan", "a,a@a.com"}, {undefined, "b@b.com"}]},
                 smtp_util:parse_rfc822_addresses(
                   <<"Jan <a,a@a.com>,b@b.com">>)),
              ?assertEqual(
                 {ok, [{"Jan", "a,a@a.com"}, {"Berend Botje", "b@b.com"}]},
                 smtp_util:parse_rfc822_addresses(
                   <<"Jan <a,a@a.com>,\"Berend Botje\" <b@b.com>">>))
      end}
    ].

combine_rfc822_addresses_test_() ->
    [
     {"One address",
      fun() ->
              ?assertEqual(
                 <<"john@doe.com">>,
                 smtp_util:combine_rfc822_addresses(
                   [{undefined, "john@doe.com"}])),
              ?assertEqual(
                 <<"John <john@doe.com>">>,
                 smtp_util:combine_rfc822_addresses(
                   [{"John", "john@doe.com"}])),
              ?assertEqual(
                 <<"\"John \\\"Foo\" <john@doe.com>">>,
                 smtp_util:combine_rfc822_addresses(
                   [{"John \"Foo", "john@doe.com"}]))
      end},

     {"Multiple addresses",
      fun() ->
              ?assertEqual(
                 <<"john@doe.com, foo@bar.com">>,
                 smtp_util:combine_rfc822_addresses(
                   [{undefined, "john@doe.com"}, {undefined, "foo@bar.com"}])),
              ?assertEqual(
                 <<"John <john@doe.com>, foo@bar.com">>,
                 smtp_util:combine_rfc822_addresses(
                   [{"John", "john@doe.com"}, {undefined, "foo@bar.com"}]))
      end}
    ].

illegal_rfc822_addresses_test_() ->
    [
     {"Nested brackets",
      fun() ->
              ?assertEqual(
                 {error,{0,smtp_rfc822_parse,["syntax error before: ","\">\""]}},
                 smtp_util:parse_rfc822_addresses("a<b<c>>"))
      end}
    ].

rfc822_addresses_roundtrip_test() ->
    Addr = <<"Jan <a,a@a.com>, Berend Botje <b@b.com>">>,
    {ok, Parsed} = smtp_util:parse_rfc822_addresses(Addr),
    ?assertEqual(Addr, smtp_util:combine_rfc822_addresses(Parsed)),
    ok.


src/gen_smtp_server.erl
%%% Copyright 2009 Andrew Thompson <andrew@hijacked.us>. All rights reserved.
%%%
%%% Redistribution and use in source and binary forms, with or without
%%% modification, are permitted provided that the following conditions are met:
%%%
%%%   1. Redistributions of source code must retain the above copyright notice,
%%%      this list of conditions and the following disclaimer.
%%%   2. Redistributions in binary form must reproduce the above copyright
%%%      notice, this list of conditions and the following disclaimer in the
%%%      documentation and/or other materials provided with the distribution.
%%%
%%% THIS SOFTWARE IS PROVIDED BY THE FREEBSD PROJECT ``AS IS'' AND ANY EXPRESS OR
%%% IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
%%% MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
%%% EVENT SHALL THE FREEBSD PROJECT OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
%%% INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
%%% (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
%%% LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
%%% ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
%%% (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
%%% SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

%% @doc A non-blocking tcp listener for SMTP connections. Based on the
%% tcp_listener module by Serge Aleynikov
%% [http://www.trapexit.org/Building_a_Non-blocking_TCP_server_using_OTP_principles]
%% @end

-module(gen_smtp_server).
-behaviour(gen_server).

-define(PORT, 2525).

%% External API
-export [start_link/3, start_link/2, start_link/1,
         start/3, start/2, start/1,
         stop/1, sessions/1].

%% gen_server callbacks
-export [init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2,
         code_change/3].


-record listener, {
          hostname :: list(),
          port :: port(),
          sessionoptions = [] :: [tuple()],
          socket :: port() | any(),
          listenoptions = [] :: [tuple()]
         }.

-type listener() :: #listener{}.

-record state, {
          listeners :: [listener()],  % Listening sockets (tcp or ssl)
          module :: atom(),
          sessions = [] :: [pid()]
         }.

-type options() ::
        [{'domain', string()} |
         {'address', {pos_integer(), non_neg_integer(),
                      non_neg_integer(), non_neg_integer()}} |
         {'port', pos_integer()} |
         {'protocol', 'tcp' | 'ssl'} |
         {'sessionoptions', [any()]}].

%% @doc Start the listener as a registered process with callback
%% module `Module' on with options `Options' linked to the calling
%% process.
%% @end
-spec start_link/3 :: (ServerName :: {'local', atom()} | {'global', any()},
                       Module :: atom(), Options :: [options()]) ->
                              {'ok', pid()} | 'ignore' | {'error', any()}.
start_link(ServerName, Module, Options) when is_list(Options) ->
	gen_server:start_link(ServerName, ?MODULE, [Module, Options], []).

%% @doc Start the listener with callback module `Module' on with
%% options `Options' linked to the calling process.
%% @end
-spec start_link/2 :: (Module :: atom(), Options :: [options()]) ->
                              {'ok', pid()} | 'ignore' | {'error', any()}.
start_link(Module, Options) when is_list(Options) ->
	gen_server:start_link(?MODULE, [Module, Options], []).

%% @doc Start the listener with callback module `Module' with default
%% options linked to the calling process.
%% @end
-spec start_link/1 :: (Module :: atom()) ->
                              {'ok', pid()} | 'ignore' | {'error', any()}.
start_link(Module) ->
	start_link(Module, [[]]).

%% @doc Start the listener as a registered process with callback
%% module `Module' with options `Options' linked to no process.
%% @end
-spec start/3 :: (ServerName :: {'local', atom()} | {'global', any()},
                  Module :: atom(), Options :: [options()]) ->
                         {'ok', pid()} | 'ignore' | {'error', any()}.
start(ServerName, Module, Options) when is_list(Options) ->
	gen_server:start(ServerName, ?MODULE, [Module, Options], []).

%% @doc Start the listener with callback module `Module' with options
%% `Options' linked to no process.
%% @end
-spec start/2 :: (Module :: atom(), Options :: [options()]) ->
                         {'ok', pid()} | 'ignore' | {'error', any()}.
start(Module, Options) when is_list(Options) ->
	gen_server:start(?MODULE, [Module, Options], []).

%% @doc Start the listener with callback module `Module' with default
%% options linked to no process.
%% @end
-spec start/1 :: (Module :: atom()) ->
                         {'ok', pid()} | 'ignore' | {'error', any()}.
start(Module) ->
	start(Module, [[]]).

%% @doc Stop the listener pid() `Pid' with reason `normal'.
-spec(stop/1 :: (Pid :: pid()) -> 'ok').
stop(Pid) ->
	gen_server:call(Pid, stop).

%% @doc Return the list of active SMTP session pids.
-spec sessions(Pid :: pid()) -> [pid()].
sessions(Pid) ->
	gen_server:call(Pid, sessions).

%% @doc
%% The gen_smtp_server is given a list of tcp listener configurations.
%% You'll typically only want to listen on one port so your options
%% will be a single-item list containing a proplist. e.g.:
%%
%% <pre>  [[{port,25},{protocol,tcp},{domain,"myserver.com"},{address,,{0,0,0,0}}]]</pre>
%%
%% By providing additional configurations the server will listen on multiple
%% ports over either tcp or ssl on any given port. e.g.:
%% <pre>
%%  [
%%    [{port,25},{protocol,tcp},{domain,"myserver.com"},{address,{0,0,0,0}}],
%%    [{port,465},{protocol,ssl},{domain,"secure.myserver.com"},{address,{0.0.0.0}}],
%%    [{port, 25},{family, inet6},{domain,"ipv6.myserver.com"},{address,"::"}]
%%  ]
%% </pre>
%% Note that the default port to listen on is `2525' because the regular SMTP
%% ports are privileged and only bindable by root. The default protocol is
%% `tcp', the default listen address is `0.0.0.0' and the default address family
%% is `inet'. Anything passed in the `sessionoptions' option, is passed through
%% to `gen_server_smtp_session'.
%% @see gen_smtp_server_session
-spec init/1 :: (Args :: list()) -> {'ok', #state{}} | {'stop', any()}.
init([Module, Configurations]) ->
	process_flag(trap_exit, true),

	DefaultConfig = [{domain, smtp_util:guess_FQDN()}, {address, {0,0,0,0}},
		{port, ?PORT}, {protocol, tcp}, {family, inet}],

    case Configurations of
        [FirstConfig|_] when is_list(FirstConfig) ->
            error_logger:info_msg("~p starting at ~p~n", [?MODULE, node()]),
            Listeners =
                [extract_listener(Config, DefaultConfig) ||
                    Config <- Configurations],

            DropPredicat = fun(R) -> element(1, R) =/= error end,
            case lists:dropwhile(DropPredicat, Listeners) of
                [] ->
                    {ok, #state{listeners = Listeners, module = Module}};
                _Else ->
                    {stop, {init, hd(Listeners)}}
            end;
        _ ->
            {stop, {init,
                    "Please start gen_smtp_server with an options "
                    "argument formatted as a list of proplists"}}
    end.

extract_listener(Config, DefaultConfig) ->
    NewConfig =
        lists:ukeymerge(1, lists:sort(Config), lists:sort(DefaultConfig)),
    Port = proplists:get_value(port, NewConfig),
    IP = proplists:get_value(address, NewConfig),
    Family = proplists:get_value(family, NewConfig),
    Hostname = proplists:get_value(domain, NewConfig),
    Protocol = proplists:get_value(protocol, NewConfig),
    SessionOptions = proplists:get_value(sessionoptions, NewConfig, []),
    ListenOptions = [binary, {ip, IP}, Family],
    case socket:listen(Protocol, Port, ListenOptions) of
        {ok, ListenSocket} -> %%Create first accepting process
            error_logger:info_msg("~p listening on ~p:~p via ~p~n",
                                  [?MODULE, IP, Port, Protocol]),
            socket:begin_inet_async(ListenSocket),
            #listener{port = socket:extract_port_from_socket(ListenSocket),
                      hostname = Hostname, sessionoptions = SessionOptions,
                      socket = ListenSocket, listenoptions = ListenOptions};
        {error, Reason} = Error ->
            error_logger:error_msg(
              "~p could not listen on ~p:~p via ~p. Error: ~p~n",
              [?MODULE, IP, Port, Protocol, Reason]),
            Error
    end.

%% @hidden
-spec handle_call(Message :: any(),
                  From :: {pid(), reference()},
                  State :: #state{}) ->
                         {'stop', 'normal', 'ok', #state{}} |
                         {'reply', any(), #state{}}.
handle_call(stop, _From, State) ->
	{stop, normal, ok, State};

handle_call(sessions, _From, State) ->
	{reply, State#state.sessions, State};

handle_call(Request, _From, State) ->
	{reply, {unknown_call, Request}, State}.

%% @hidden
-spec handle_cast(Message :: any(), State :: #state{}) -> {'noreply', #state{}}.
handle_cast(_Msg, State) ->
	{noreply, State}.

%% @hidden
-spec handle_info(Message :: any(), State :: #state{}) ->
                         {'noreply', #state{}} | {'stop', any(), #state{}}.
handle_info({inet_async, ListenPort,_, {ok, ClientAcceptSocket}},
	#state{module = Module,
           listeners = Listeners,
           sessions = CurSessions} = State) ->
	try
		% find this ListenPort in our listeners.
		[Listener] = lists:flatten([case L#listener.port of
					ListenPort -> L;
					_ -> []
				end || L <- Listeners]),
		{ok, ClientSocket} =
            socket:handle_inet_async(
              Listener#listener.socket,
              ClientAcceptSocket,
              Listener#listener.listenoptions),

		%% New client connected
		% io:format("new client connection.~n", []),
        Startup =
            gen_smtp_server_session:start(
              ClientSocket, Module,
              [{hostname, Listener#listener.hostname},
               {sessioncount, length(CurSessions) + 1} |
               Listener#listener.sessionoptions]),

		Sessions =
            case Startup of
                {ok, Pid} ->
                    link(Pid),
                    socket:controlling_process(ClientSocket, Pid),
                    CurSessions ++[Pid];
                _Other ->
                    CurSessions
            end,
		{noreply, State#state{sessions = Sessions}}
	catch _:Error ->
            error_logger:error_msg("Error in socket acceptor: ~p.~n", [Error]),
            {noreply, State}
	end;
handle_info({'EXIT', From, Reason}, State) ->
	case lists:member(From, State#state.sessions) of
		true ->
			{noreply,
             State#state{sessions = lists:delete(From, State#state.sessions)}};
		false ->
			io:format("process ~p exited with reason ~p~n", [From, Reason]),
			{noreply, State}
	end;
handle_info({inet_async, ListenSocket, _, {error, econnaborted}}, State) ->
	io:format("Client terminated connection with econnaborted~n"),
	socket:begin_inet_async(ListenSocket),
	{noreply, State};
handle_info({inet_async, _ListenSocket,_, Error}, State) ->
	error_logger:error_msg("Error in socket acceptor: ~p.~n", [Error]),
	{stop, Error, State};
handle_info(_Info, State) ->
	{noreply, State}.

%% @hidden
-spec terminate(Reason :: any(), State :: #state{}) -> 'ok'.
terminate(Reason, State) ->
	io:format("Terminating due to ~p~n", [Reason]),
	lists:foreach(
      fun(#listener{socket=S}) -> catch socket:close(S) end,
      State#state.listeners),
	ok.

%% @hidden
-spec code_change(OldVsn :: any(), State :: #state{}, Extra :: any()) ->
                         {'ok', #state{}}.
code_change(_OldVsn, State, _Extra) ->
	{ok, State}.
